include/rmm/mr/device/statistics_resource_adaptor.hpp

```diff
@@ -186,8 +186,8 @@ class statistics_resource_adaptor final : public device_memory_resource {
   std::pair<counter, counter> push_counters()
   {
     write_lock_t lock(mtx_);
-    auto ret = counter_stack_.top();
-    counter_stack_.push(std::make_pair(counter{}, counter{}));
+    auto ret = counter_stack_.empty() ? std::make_pair(counter{}, counter{}) : counter_stack_.top();
+    counter_stack_.push({counter{}, counter{}});
```

Comment:
*question*: While this fixes the issue. It does (internally) invalidate the docstring (in the ctor the return value from counter_stack_ is the newly pushed set of counters, though it doesn't matter because it's ignored).

There's also some action-at-distance in that it's not fully obvious that the reason for the `push_counters` in the ctor is that one needs to maintain the invariant that the counter stack always has at least one entry.

Can we instead just initialize the counter stack appropriately at construction time, then the necessary invariant for the stack is always true:
```diff
diff --git a/include/rmm/mr/device/statistics_resource_adaptor.hpp b/include/rmm/mr/device/statistics_resource_adaptor.hpp
index c8906afd..40a247ff 100644
--- a/include/rmm/mr/device/statistics_resource_adaptor.hpp
+++ b/include/rmm/mr/device/statistics_resource_adaptor.hpp
@@ -122,8 +122,6 @@ class statistics_resource_adaptor final : public device_memory_resource {
   statistics_resource_adaptor(Upstream* upstream) : upstream_{upstream}
   {
     RMM_EXPECTS(nullptr != upstream, "Unexpected null upstream resource pointer.");
-    // Initially, we push a single counter pair on the stack
-    push_counters();
   }
 
   statistics_resource_adaptor()                                              = delete;
@@ -196,6 +194,7 @@ class statistics_resource_adaptor final : public device_memory_resource {
    *
    * @return top pair of counters <bytes, allocations> from the stack _before_
    * the pop
+   * @throws std::out_of_range if the counter stack has fewer than two entries.
    */
   std::pair<counter, counter> pop_counters()
   {
@@ -275,7 +274,8 @@ class statistics_resource_adaptor final : public device_memory_resource {
   }
 
   // Stack of counter pairs <bytes, allocations>
-  std::stack<std::pair<counter, counter>> counter_stack_;
+  // Invariant: the stack always contains at least one entry
+  std::stack<std::pair<counter, counter>> counter_stack_{{std::make_pair(counter{}, counter{})}};
   std::shared_mutex mutable mtx_;  // mutex for thread safe access to allocations_
   Upstream* upstream_;             // the upstream resource used for satisfying allocation requests
 };
```
WDYT?

Reply:
LGTM

Reply:
Yeah, I think this is a good idea. I started something similar but got lazy. 

I find it a little bit incongruous to throw std::out_of_range here when std::stack (and the underlying container it is built on) do not throw when you `pop()`/`pop_back` an empty container: it is just UB.

More importantly, I need to ask: is `pop_counters()` likely to be called from a destructor or memory deallocation? If so, it must be `noexcept`. It *will* be called at the point where destruction happens -- when leaving scope. If it is implemented using RAII, then it is called in a dtor, and so must be `noexcept`.  Thoughts?

Reply:
> More importantly, I need to ask: is pop_counters() likely to be called from a destructor or memory deallocation? If so, it must be noexcept. It will be called at the point where destruction happens -- when leaving scope. If it is implemented using RAII, then it is called in a dtor, and so must be noexcept. Thoughts?

`pop_counters` is not called by destructors in the current implementation, and the usage right now is of the form:

```
auto some_func(...)
{
   mr.push_counters();
   make_calls_that_allocate/deallocate();
   stats = mr.pop_counters();
   // record stats somewhere
}
```

Which I think is safe. If one wanted to wrap that push/pop/record idiom in an RAII object, then yes, we would need to change `pop_counters()` to `noexcept` and (instead of `throw`ing) `debug_assert` if the stack invariant is not maintained due to mismatched push/pop pairs.

Reply:
Seems to me this should be a debug assert and not a throw. 
 * This would allow us to add RAII functionality, and make it easier to use in an NVTX RAII object.
 * The invariant is documented and maintained by all correct usage of the class.
 * The underlying C++ containers don't throw, they document it as UB

Reply:
Argh, `std::shared_mutex::lock()` can throw `std::system_error`. Rare, but making these functions `noexcept` would be wrong...